ddtrace/internal/writer.py

```diff
@@ -355,21 +359,27 @@ def recreate(self):
         )
 
     def _put(self, data, headers):
-        conn = get_connection(self.agent_url, self._timeout)
-
+        # type: (bytes, Dict[str, str]) -> Response
         with StopWatch() as sw:
+            if self._conn is None:
+                self._conn = get_connection(self.agent_url, self._timeout)
+
             try:
-                conn.request("PUT", self._endpoint, data, headers)
-                resp = compat.get_connection_response(conn)
+                self._conn.request("PUT", self._endpoint, data, headers)
+            except Exception:
```

Comment:
@majorgreys: Is there a particular exception that will be raised that is expected for when keep-alive doesn't work?

Reply:
Looks like potentially [`RemoteDisconnected`](https://docs.python.org/3/library/http.client.html#http.client.RemoteDisconnected) but the docs say

> Raised by [HTTPConnection.getresponse()](https://docs.python.org/3/library/http.client.html#http.client.HTTPConnection.getresponse) when the attempt to read the response results in no data read from the connection, indicating that the remote end has closed the connection.

which I just noticed that we don't include the call to `getresponse` in the exception handling here, so that change should be made.

As for handling the exceptions particularly [`HTTPException`](https://docs.python.org/3/library/http.client.html#http.client.HTTPException) is the base for all other http related exceptions and is handled by tenacity which invokes this method so I don't think there's a need to be specific here.